dfp/decimal/decimal

```diff
@@ -180,6 +180,286 @@ public:
+    dfp_num_get<_CharT, _InIter>::
+    do_get(iter_type __beg, iter_type __end, ios_base& __io,
+           ios_base::iostate& __err, decimal::decimal32& __v) const
+    {
+      string __xtrc;
+      __xtrc.reserve(32);
+      __beg = this->_M_extract_float(__beg, __end, __io, __err, __xtrc);
+      std::__convert_to_v(__xtrc.c_str(), __v, __err, locale::facet::_S_get_c_locale());
+      if (__beg == __end)
+        __err |= ios_base::eofbit;
+      return __beg;
+    }
+
+  template<typename _CharT, typename _InIter>
+    _InIter
+    dfp_num_get<_CharT, _InIter>::
+    do_get(iter_type __beg, iter_type __end, ios_base& __io,
+           ios_base::iostate& __err, decimal::decimal64& __v) const
+    {
+      string __xtrc;
+      __xtrc.reserve(32);
+      __beg = this->_M_extract_float(__beg, __end, __io, __err, __xtrc);
+      std::__convert_to_v(__xtrc.c_str(), __v, __err, locale::facet::_S_get_c_locale());
+      if (__beg == __end)
+        __err |= ios_base::eofbit;
+      return __beg;
+    }
+
+  template<typename _CharT, typename _InIter>
+    _InIter
+    dfp_num_get<_CharT, _InIter>::
+    do_get(iter_type __beg, iter_type __end, ios_base& __io,
+           ios_base::iostate& __err, decimal::decimal128& __v) const
+    {
+      string __xtrc;
+      __xtrc.reserve(32);
+      __beg = this->_M_extract_float(__beg, __end, __io, __err, __xtrc);
+      std::__convert_to_v(__xtrc.c_str(), __v, __err, locale::facet::_S_get_c_locale());
+      if (__beg == __end)
+        __err |= ios_base::eofbit;
+      return __beg;
+    }
+
+  template <class charT, class traits = char_traits<charT>>
+  class dfp_basic_istream;
+
+  namespace decimal
+  {
+    template <class charT, class traits>
+    std::dfp_basic_istream<charT, traits> &
+      operator>>(std::dfp_basic_istream<charT, traits> & is, decimal32 & d);
+  }
+
+  template <class charT, class traits>
+  class dfp_basic_istream : public basic_istream<charT, traits>
+  {
+
+    template<class X, class Y>
+    friend dfp_basic_istream<X, Y> &
+      decimal::operator>>(dfp_basic_istream<X, Y> & is, decimal::decimal32 & d);
+
+    template<class X, class Y, class T>
+    friend basic_istream<X, Y> &
+      decimal::operator>>(basic_istream<X, Y> & is, T & d);
+
+    template<typename _ValueT>
+    dfp_basic_istream<charT, traits> &
+    _M_extract(_ValueT& d);
+  };
+
+  template<typename _CharT, typename _Traits>
+    template<typename _ValueT>
+      dfp_basic_istream<_CharT, _Traits>&
+      dfp_basic_istream<_CharT, _Traits>::
+      _M_extract(_ValueT& __v)
+      {
+        typename basic_istream<_CharT, _Traits>::sentry __cerb(*this, false);
+        if (__cerb)
+          {
+            ios_base::iostate __err = ios_base::goodbit;
+            __try
+              {
+                locale dfp_locale {this->_M_ios_locale, new dfp_num_get<_CharT>()};
+                auto& facet {use_facet<dfp_num_get<_CharT>>(dfp_locale)};
+                facet.get(*this, 0, *this, __err, __v);
```

Comment:
Why bother to create a locale using the facet, and then extract the facet again (with expensive RTTI checks to make sure it's present and has the right type)? You just need the facet.

```c++
dfp_num_get<_CharT> ng;
ng.get(*this, 0, *this, __err, __v);
```